lib/Model/ServerStorageDevices.php

```diff
@@ -1,231 +1,50 @@
-    protected static $swaggerModelName = 'Server_storage_devices';
-
-    /**
-      * Array of property to type mappings. Used for (de)serialization
-      * @var string[]
-      */
-    protected static $swaggerTypes = [
-        'storage_device' => '\Upcloud\ApiClient\Model\StorageDevice[]'
-    ];
-
-    /**
-      * Array of property to format mappings. Used for (de)serialization
-      * @var string[]
-      */
-    protected static $swaggerFormats = [
-        'storage_device' => null
-    ];
-
-    public static function swaggerTypes()
-    {
-        return self::$swaggerTypes;
-    }
-
-    public static function swaggerFormats()
-    {
-        return self::$swaggerFormats;
-    }
-
     /**
-     * Array of attributes where the key is the local name, and the value is the original name
-     * @var string[]
+     * @var StorageDevice[]|null
      */
-    protected static $attributeMap = [
-        'storage_device' => 'storage_device'
-    ];
-
-
-    /**
-     * Array of attributes to setter functions (for deserialization of responses)
-     * @var string[]
-     */
-    protected static $setters = [
-        'storage_device' => 'setStorageDevice'
-    ];
-
-
-    /**
-     * Array of attributes to getter functions (for serialization of requests)
-     * @var string[]
-     */
-    protected static $getters = [
-        'storage_device' => 'getStorageDevice'
-    ];
-
-    public static function attributeMap()
-    {
-        return self::$attributeMap;
-    }
-
-    public static function setters()
-    {
-        return self::$setters;
-    }
-
-    public static function getters()
-    {
-        return self::$getters;
-    }
-
-    
-
-    
-
-    /**
-     * Associative array for storing property values
-     * @var mixed[]
-     */
-    protected $container = [];
+    private $storageDevice;
 
     /**
      * Constructor
-     * @param mixed[] $data Associated array of property values initializing the model
+     * @param mixed[] $data
      */
     public function __construct(array $data = null)
     {
-        $this->container['storage_device'] = isset($data['storage_device']) ? $data['storage_device'] : null;
+        $this->setStorageDevice($data['storage_device'] ?? null);
     }
 
     /**
-     * show all the invalid properties with reasons.
-     *
-     * @return array invalid properties with reasons
+     * @return StorageDevice[]|null
      */
-    public function listInvalidProperties()
+    public function getStorageDevice(): ?array
     {
-        $invalid_properties = [];
-
-        return $invalid_properties;
+        return $this->storageDevice;
     }
 
     /**
-     * validate all the properties in the model
-     * return true if all passed
-     *
-     * @return bool True if all properties are valid
+     * @param StorageDevice[]|null $storageDevice
+     * @return ServerStorageDevices
      */
-    public function valid()
+    public function setStorageDevice(?array $storageDevice): ServerStorageDevices
```

Comment:
Since the function is named as `setStorageDevice()` I would accept a single storage device object as param not a list. (ReadMe example passes a storage device object to this function)

Reply:
The function named this way because it named according to names in the source API (e.g. https://i.imgur.com/MbQCQNv.png). There we have key(singular) and value, which is an array. 

Regarding ReadMe, actually, this example hasn't been edited and was incorrect even before. 

Currently function `setStorageDevice ` accepts an array of arrays. We can adjust it, so it would accept an array of objects if you'd like. But it can't accept a single object.

Reply:
Yeah, the UpCloud API is a bit bad in that regard. Thoug we should not copy bad practices from there :joy: Unless the library is coded so that you need to create that nesting yourself (i would prefer so that the library would do that for you).

(FYI, the `"storage_devices": { storage_device: [] }` nesting in API is because of old XML support thingies)

Reply:
@Darep @PopoSensei We'll take a look at this issue, and give your our feedback regarding what we can do after we investigate it.

Reply:
@Darep @PopoSensei Could you guys give a clarification on the changes you want us to make? 

The first comment by @PopoSensei was regarding the singular function name `setStorageDevice()`, and you suggested that functions that receive arrays as parameters should have plural names. Then @Darep mentioned the nesting issue.

Please, let us know the exact changes we need to make. Thanks!

Reply:
I agree that functions that receive arrays as parameters should have plural names :) So I would suggest `setStorageDevices()` here. Probably nothing else required right now.

You can disregard my comment about the nesting; just wanted to let you know the legacy reasons behind it, and that I personally think we should avoid introducing or exposing that annoying nesting as much as possible :blush:

Reply:
@PopoSensei @Darep Regarding plural names: 

Since we have such nestings (e.g. `"storage_devices": { storage_device: [] }`), objects instances are made in accordance to the API. Because of that, we have such code blocks: 
`$storageDevice = $server->getStorageDevices()->getStorageDevice();`

If we try to rename singular function names to plural, we'll have something like:
`$storageDevice = $server->getStorageDevices()->getStorageDevices();`

We think that can be confusing and inconvenient for users.

Let us know if this repetition would be okay for you.

Reply:
Hmmm... Seems that the example storage template UUID was wrong. And once I changed that to `"01000000-0000-4000-8000-000030060200"` (copied from python-api) everything went smoothly

Reply:
@PopoSensei Yeah, the UUID's in examples can be wrong since we didn't actually edit the docs. 

Please, let us know your thoughts on our message above (regarding plural function names).

Reply:
Was `$storageDevice = $server->getStorageDevices()->getStorageDevice();` how it was previously too? Yeah, I vote for keeping that then :P 

With regards to my comments about nesting: optimal in my eyes would be `$storageDevice = $server->getStorageDevices();` so then we would abstract the nesting away from the users, but we probably shouldn't do that change now if it was like that before already (was it?)

Reply:
@Darep Yes, it was like it before. We haven't changed it, because we've tried to keep support for previous versions. 
What we did was remove lots of duplicate code and add some fixes, but the general logic is the same since we need to keep backward compatibility.

Reply:
Sounds good to me then! Sorry for the confusion and back'n'forth :)

P.S. Hopefully we can make a change in the future where we remove that nesting in code 😊  Let's keep that in mind

Reply:
Sorry seems I ran that with the old version and not the PR version. This still returns

```
PHP Fatal error:  Uncaught TypeError: Argument 1 passed to Upcloud\ApiClient\Model\ServerStorageDevices::setStorageDevice() must be of the type array or null, object given, called in /Users/trdls/upcloud-dev/apis/thirdparty-repos/devimasolutions/test-php-api/main.php on line 30 and defined in /Users/trdls/upcloud-dev/apis/thirdparty-repos/devimasolutions/test-php-api/vendor/upcloudltd/upcloud-php-api/lib/Model/ServerStorageDevices.php:35
```

So there is something very different using this updated version.

Reply:
@PopoSensei We've just pushed a new commit [81cae71](https://github.com/DevimaSolutions/upcloud-php-api-fork/commit/ad639661271fc2313ede5137f9013d5dbc6984b0), where we updated this example in docs, and now you can pass object like so https://i.imgur.com/BSeLPyo.png.

Reply:
But that changes the way the API is used. So if our users have their code like our example is, they would need to change their usage of storage and now it seems that other resources as well or am I wrong on that?

I know that I said I don't like the `setStorageDevice(array)` since it is confusing, but I believe we need to go with what we have for now. Also just post code snippets here. No need to post pictures to imgur.

Reply:
Talked about this in detail. It seems the documented example was working because of PHP:s loose typing. Which is in this PR fixed to strict typing to help maintain the API. So will resolve this conversation now :)